src/forest.cpp

```diff
@@ -131,6 +128,74 @@ namespace libsemigroups {
     }
   }
 
+  void Forest::throw_if_not_acyclic(node_type new_node,
+                                    node_type new_parent) const {
+    size_t const        not_yet_seen = number_of_nodes();
+    std::vector<size_t> seen(number_of_nodes(), not_yet_seen);
+    size_t              current = 0;
+
+    for (Forest::node_type m = 0; m != number_of_nodes(); ++m) {
+      auto      n = m;
+      node_type p;
+      size_t    length = 0;
+      while (n != UNDEFINED && seen[n] == not_yet_seen) {
+        seen[n] = current;
+        p       = n;
+        n       = parent(n);
+        length++;
+      }
+      if (n != UNDEFINED && seen[n] == current) {
+        std::string cycle;
+        if (length == 1) {
+          cycle = fmt::format("loop [{}]", n);
+        } else if (length == 2) {
+          cycle = fmt::format("cycle [{}, {}]", n, parent(n));
+        } else {
+          cycle = fmt::format(
+              "cycle [{}, {}, ..., {}] (length {})", n, parent(n), p, length);
+        }
+        if (new_node != UNDEFINED && new_parent != UNDEFINED) {
+          LIBSEMIGROUPS_EXCEPTION("defining the parent of node {} to be {} "
+                                  "creates a {}",
+                                  new_node,
+                                  new_parent,
+                                  cycle);
+        } else {
+          LIBSEMIGROUPS_EXCEPTION("the Forest object contains the {}"
+                                  " and is invalid",
+                                  cycle);
+        }
+      }
+      current++;
+    }
+  }
```

Comment:
If this is performance critical, I think we can improve the complexity from $\mathcal{O}(n^2)$ to $\mathcal{O}(n + e)$ by doing a topological sort here instead. A directed graph is acyclic if and only if it can be topologically sorted.

If detecting the location of the cycles is important too, [this StackOverflow post](https://stackoverflow.com/questions/583876/how-do-i-check-if-a-directed-graph-is-acyclic) tells me that [Tarjan's Algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm) is similar to a topological sort and can detect where the cycles are.

Reply:
I'm pretty sure the complexity of the implementation is $O(n)$ where $n$ is the number of nodes.

Reply:
Yeah. I incorrectly interpreted this as:
```
for each node:
  follow the parent until you get to the root or find a repeat
```
which I think is almost what happens, except we might stop early if we see a node we've already explored.

I don't quite understand why `seen` needs to be a vector of integers rather than a vector of bools, but that isn't really important.